runtime/major_gc.c

```diff
@@ -458,6 +423,57 @@ void caml_add_to_orphaned_ephe_list(struct caml_ephe_info* ephe_info)
     ephe_info->must_sweep_ephe = 0;
     atomic_fetch_add_verify_ge0(&num_domains_to_ephe_sweep, -1);
   }
+  CAMLassert (ephe_info->live == 0);
+  CAMLassert (ephe_info->todo == 0);
+}
+
+void caml_orphan_finalisers (caml_domain_state* domain_state)
+{
+  struct caml_final_info* f = domain_state->final_info;
+
+  if (f->todo_head != NULL || f->first.size != 0 || f->last.size != 0) {
+    /* have some final structures */
+    atomic_fetch_add(&num_domains_orphaning_finalisers, +1);
+    if (caml_gc_phase != Phase_sweep_and_mark_main) {
+      /* Force a major GC cycle to simplify constraints for orphaning
+         finalisers. See note attached to the declaration of
+         [num_domains_orphaning_finalisers] variable in major_gc.c */
+      caml_finish_major_cycle();
+    }
+    CAMLassert(caml_gc_phase == Phase_sweep_and_mark_main);
+    CAMLassert (!f->updated_first);
+    CAMLassert (!f->updated_last);
+
+    /* Add the finalisers to [orph_structs] */
+    caml_plat_lock(&orphaned_lock);
+    f->next = orph_structs.final_info;
+    orph_structs.final_info = f;
+    caml_plat_unlock(&orphaned_lock);
+
+    /* Create a dummy final info */
+    domain_state->final_info = caml_alloc_final_info();
+    f = domain_state->final_info;
```

Comment:
This is a bug fix. 

Without updating `f` here, the update to `f->updated_first` and `f->updated_last` in the rest of the function will be on the orphaned final info and not on the newly allocated final info. As a result, the next time `caml_orphan_finalisers` is called (note that `domain_terminate` calls `caml_orphan_ephemerons` in a loop), we will check the `updated_*` fields of the dummy final info allocated in the previous call to `caml_orphan_finalisers`. Those fields are 0 in the newly allocated final info and hence, we will decrement `num_domains_to_final_update*` fields again, leading to inconsistent state.

Reply:
This might be a case where the C idiom

```c
f = domain_state->final_info = caml_alloc_final_info();
```

is warranted, as morally we are updating both `final_info` and its local view `f` at the same time.